### DISCUSSION TITLE
Welcome to QingStation Discussions!

### DISCUSSION THREAD

**Discussion by majianjia** [majianjia/QingStation#1] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Antiath** [comment]
Hi
Just read rapidly your documentation.  Very interesting.

I also tried  to build an ultrasonic anemometer  but not the compact version, I choose the one with 4 arms to have as much distance as possible and some timing margin.   Other than that, I run in almost every single issues you mentionned.  At the moment the prototype is unfinished and in standby ( my last pcb design was an expensive and destructive failure so I pretty much lost my motivation after that) . I do have a "working" design on perfboard. For now, the basis of the design is mostly inspired by documentation found on this site https://soldernerd.com/ (but the author deleted the pages I think ...).  

On the emitter part, I feed the output of a mosfet driver directly into the tranducer ( 0-12V). First 12.5 µs half wave (40khZ) goes into the transducer like this : 12V on pin 1, 0 on pin 2. The next half wave is reversed : 0V on pin 1 and 12V on pin 2. And so on. I found that the response from the transducer was much stronger with this method. It also removes the need for negative voltages  on the driver side. 

On the receiving part, everything goes through a two stage  multiplexing arrangement  (designed to help minimize crosstalk) . This idea was taken from this project : https://hackaday.com/2013/08/21/ultrasonic-anemometer-for-an-absurdly-accurate-weather-station/
After that, I use a two stage amplifier with band pass filtering and also a comparator to detect zero crossings. The zerocrossings allow me to trigger the ADC and get a precise time base.  I then collect the samples and compare them to a reference sample (recorded at 0 m/s speed) with a simple cross correlation, a.k.a matched filtering.  I get the phase delay  from that. Results are really noisy. Not unreadable but barely useful.

The main issue with this design is that the echoes are so spread out that it becomes difficult to extract a good reading out of the noise. 
Add to this additionnal issues that I won't go into to keep it short. 
There is a good chance that going to higher  frequencies would resolve this, because the pulses would be more compact. But then it becomes defficult to design the electronics around that. You would need a faster driver ( not too difficult but still...)  but mainly a really fast ADC.  I tested my design with a Teensy 4 board and used it's internal ADC. Amazing piece of tech with a blazing fast microcontroller.  And the 40Khz signal is already reaching the limits of what the ADC can do.  

Good luck with your project :)

**Reply by majianjia** [reply]
I used JLCPCB which is quite cheap inc postage to the UK. There were also some mistakes in my first PCB but I managed to do some basic tests to validate my driver and analog design.

A stronger driver will definitely bring better modulation ability and a cleaner signal. But I more concerned about the size so I used the old trick, an RS232 interfacing IC to drive these transducers. Only have 10Vpp unlike yours have 24Vpp. 

I don't have good experience with analog circuit designs, so what I did is to keep the driver and receiver(opamp) as simple as possible. Then use signal processing to do all the zero-crossing detection and pattern matching. End up with a lot more work to do compared to hardware solutions.  The ADC start sampling (continue sampling) right after the timer sends PWM to the driver (triggered by internal hardware event), to avoid any jitter caused by software.  Only ensure this can make the zero-crossing interpolation meaningful. The accuracy increases very much. 

The MCU I has a much lower performance than the Teensy 4. I think you just need to change the way that the ADC samples the data. 

Matching the pattern is quite difficult indeed. Most of my time is working on it. 
Besides some misunderstanding that I mention in the log, I now know more details about modulated pulse and noise, but the current hardware is not capable to do. Modulated pulse is the way to do it, and pulse compression is the key. I saw a paper use 5 V for regular excitation then use 2 pulses 12V for rapid change of phase. It produces a nearly perfect phase shift. As long as there is a good short phase shift moment, the pulse compression will provide a very good result. 

I recommend reading Lau's post, very informative. 
https://www.dl1glh.de/ultrasonic-anemometer.html#advancement

**Comment by kebasaa** [comment]
Hi, this is an amazing project. I've been looking for something like this for a long time. I was just wondering: Do you think it would be possible to create a board for the sonic anemometer that can connect to a device like the ESP32 through i2c or SPI? Then every used could program it themselves. Specifically, I'm interested in creating a 3D sonic anemometer to obtain the 3D wind direction, and to sample it at 20Hz or more (It would have to include a compass and inclination sensor of course). This would answer a lot of interesting scientific questions... How could that be built?

**Reply by majianjia** [reply]
Hi, 
I am developing the second gen of QingStation. 
At the moment, it can log measurements through its console but won't take command from other devices. 
I dont have plan to develop configurations.
